lib/fetching/content-services/comments-content-service.js

```diff
@@ -0,0 +1,170 @@
+var request = require('request');
+var ContentService = require('../content-service');
+var util = require('util');
+var logger = require('../../logger');
+var Report = require('../../../models/report');
+var config = require('../../../config/secrets');
+var { completeUrl, authenticate, getJWTToken } = require('../../comments');
+
+var CommentContentService = function(options) {
+    this.fetchType = 'pull';
+    this.interval = 10000;
+
+    const  { count, lastFetchDate, lastAuthoredDate } = config.get().comments;
+    this._count = count;
+    if (lastFetchDate) {
+        this._lastFetchDate = new Date(lastFetchDate);
+        this._lastAuthoredDate = new Date(lastAuthoredDate);
+    }
+
+    ContentService.call(this, options)
+}
+
+CommentContentService.prototype._doFetch = function(_, callback) {
+    const commentUrl = completeUrl('comment');
+    const options = {
+        body: {
+            count: this._count
+        },
+        json: true
+    };
+
+    if (this._lastFetchDate && this._lastAuthoredDate) {
+        options.body = {
+            ...options.body,
+            acqAfter: this._lastFetchDate.getTime(),
+            authAfter: this._lastAuthoredDate.getTime(),
+        }
+    }
+
+    this._sendRequest(commentUrl, options, callback);
+}
+
+CommentContentService.prototype._sendRequest = function(url, opts, callback) {
+    if (typeof getJWTToken() !== 'string') {
+        authenticate((err) => {
+            if (err) {
+                this.emit('error', err);
+                return callback([]);
+            }
+            this._sendRequest(url, opts, callback);
+        });
+        return;
+    }
+
+    opts.auth = {
+        bearer: getJWTToken(),
+    }
+    request(url, opts, (err, res, body) => {
+        if (err) {
+            this.emit('error', new Error('HTTP error: ' + err.message));
+            return callback([]);
+        } else if (res.statusCode === 401) {
+            jwtToken = null;
+            this._sendRequest(url, opts, callback);
+            return;
+        } else if (res.statusCode !== 200) {
+            this.emit('error', new Error.HTTP(res.statusCode));
+            return callback([]);
+        }
+
+        var comments = body;
+        try {
+            if (typeof comments === 'string') {
+                comments = JSON.parse(comments);
+            }
+            if (!Array.isArray(comments)) {
+                this.emit('error', new Error('Wrong data'));
+                return callback([]);
+            }
+        } catch (e) {
+            this.emit('error', new Error('Parse error: ' + e.message));
+            return callback([]);
+        }
+
+        var remaining = comments.length;
+        var reports = [];
+
+        if (comments.length === 0) {
+            return callback([]);
+        }
+
+        this.cache = {};
+
+        const { acquiredAt, timestamp } = comments[comments.length - 1];
+
+        // TODO rewrite with async/await to avoid callback hell
```

Comment:
😁 

(It's pretty straightforward to simplify this if you wanna take a swing at it, happy to send a quick example I have. Not blocking, just nice.)

Reply:
Code rewritten to use async/await :)